### PR TITLE
Benchmark templates

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120275
* __->__ #118880

Adding support for benchmarking templates in `benchmark_fusion`

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/triton_heuristics.py [hunk 1]
```diff
@@ -498,7 +498,7 @@ def benchmark_all_configs(self, *args, **kwargs):
             log.debug("Benchmark all input configs for %s, get:", self.fn.__name__)
             for k, v in timings.items():
                 log.debug(
-                    "%s: %f, nreg %d, nspill %d, #shared-mem %d",
+                    "%s: %f, nreg %d, nspill %d, #shared-mem %s",
```
Comment (shunting314) [inline torch/_inductor/triton_heuristics.py:501]: Any reason for this change?

File: torch/_inductor/codegen/triton.py [hunk 2]
```diff
@@ -3522,9 +3522,13 @@ def define_kernel(self, src_code, node_schedule):
...
+    def codegen_template(
+        self, template_node, epilogue_nodes, only_gen_src_code=False
+    ) -> Optional[str]:
         """
         Codegen a triton template
+
+        If `only_gen_src_code` the src code will be returned instead of codegen'd into the wrapper
         """
         _, (numel, rnumel) = template_node.group
         assert rnumel == 1
```
Comment (shunting314) [inline torch/_inductor/codegen/triton.py:3534]: mark_run will add memory allocation lines to the wrapper, we should skip it when we are do benchmarking.

Comment (shunting314) [reply torch/_inductor/codegen/triton.py:3534]: Not sure if it's GH bug, I can not attach this comment to line 3538 where `node.mark_run()` is called.

File: torch/_inductor/triton_heuristics.py (hunk 1 above)
Comment (eellison) [reply torch/_inductor/triton_heuristics.py:501]: at least on my machine, shared memory is sometimes None for some reason. %s will jsut convert input to string regardless of type

File: torch/_inductor/codegen/triton.py (hunk 2 above)
Comment (eellison) [reply torch/_inductor/codegen/triton.py:3534]: Good catch - addressed and extended test so it would have caught

**Review by shunting314 — APPROVED** [review]

**Pull request closed without merging**